x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/movingPercentiles/MovingPercentilesPipelineAggregator.java

```diff
@@ -0,0 +1,254 @@
+                                        internalHDRPercentiles.keyed(),
+                                        internalHDRPercentiles.formatter());
+        }
+        throw buildResolveError(agg, aggPathsList, propertyValue, "percentiles");
+    }
+
+    private TDigestState resolveTDigestBucketValue(MultiBucketsAggregation agg,
+                                                   InternalMultiBucketAggregation.InternalBucket bucket,
+                                                   String aggPath) {
+        List<String> aggPathsList = AggregationPath.parse(aggPath).getPathElementsAsStringList();
+        Object propertyValue = bucket.getProperty(agg.getName(), aggPathsList);
+        if (propertyValue == null || (propertyValue instanceof InternalTDigestPercentiles) == false) {
+            throw buildResolveError(agg, aggPathsList, propertyValue, "TDigest");
+        }
+        return ((InternalTDigestPercentiles) propertyValue).getState();
+    }
+
+    private DoubleHistogram resolveHDRBucketValue(MultiBucketsAggregation agg,
+                                                  InternalMultiBucketAggregation.InternalBucket bucket,
+                                                  String aggPath) {
+        List<String> aggPathsList = AggregationPath.parse(aggPath).getPathElementsAsStringList();
+        Object propertyValue = bucket.getProperty(agg.getName(), aggPathsList);
+        if (propertyValue == null || (propertyValue instanceof InternalHDRPercentiles) == false) {
+            throw buildResolveError(agg, aggPathsList, propertyValue, "HDR");
+        }
+        return ((InternalHDRPercentiles) propertyValue).getState();
+    }
+
+    private IllegalArgumentException buildResolveError(MultiBucketsAggregation agg, List<String> aggPathsList,
+                                                       Object propertyValue, String method) {
+        if (propertyValue == null) {
+            return new IllegalArgumentException(AbstractPipelineAggregationBuilder.BUCKETS_PATH_FIELD.getPreferredName()
+                + " must reference a " + method + " percentile aggregation");
+        } else {
+            String currentAggName;
+            if (aggPathsList.isEmpty()) {
+                currentAggName = agg.getName();
+            } else {
+                currentAggName = aggPathsList.get(0);
+            }
+            return new IllegalArgumentException(AbstractPipelineAggregationBuilder.BUCKETS_PATH_FIELD.getPreferredName()
+                + " must reference a " + method + " percentiles aggregation, got: ["
+                + propertyValue.getClass().getSimpleName() + "] at aggregation [" + currentAggName + "]");
+        }
+    }
+
+    private int clamp(int index, int length) {
+        if (index < 0) {
+            return 0;
+        }
+        if (index > length) {
+            return length;
+        }
+        return index;
+    }
+
+    // TODO: replace this with the PercentilesConfig that's used by the percentiles builder.
```

Comment:
I think it'd be a fairly small change to plumb the parent builder into the ctor of the pipeline aggregator. A fine thing to do in a follow up though because it'd touch a bunch of file mechanically.